### PR TITLE
#11403: SubMesh Support + Porting/Stamping T3K Tests to Galaxy

### PR DESCRIPTION
**Pull request by cfjchu**
### Ticket
[Link to Github Issue
](https://github.com/tenstorrent/tt-metal/issues/11403)

### Problem description

We want to add support to instantiate submeshes on our mesh. Key reasons:

1. Port over additional T3000 tests onto galaxy. There is Infrastructure support is needed to specify submeshes at arbitrary offsets in the SystemMesh.
2. Scale throughput performance by stamping-out / “tiling” our submesh on our full-mesh. 

### What's changed

1. **T3000 Tests to TG**: Supporting changes to enable all T3000 tests to be ported to TG
2. **SubMesh Support:**: Enable a user to instanatiate a submesh (MeshDevice) from a mesh. This also specializes the MeshDevice type: {RowMajor, Ring, Line}. 
3. **SubMesh Stamping**: "Stamp-out"/tile the submesh across TG. llama is added as an example.

### Example API Usage:

For example, create two ring 2x2 on T3000 machines and execute two ring all-gather on 2x2 ring.

```python
def run_model(submesh):
    full_tensor = torch.ones((1, 1, 32, 32 * submesh.get_num_devices()), dtype=torch.bfloat16)
    for i in range(submesh.get_num_devices()):
        full_tensor[..., i * 32 : (i + 1) * 32] = i

    ttnn_tensor = ttnn.from_torch(full_tensor, mesh_mapper=ShardTensorToMesh(submesh, dim=3))
    ttnn_tensor = ttnn.to_device(ttnn_tensor, submesh)
    ttnn_tensor = ttnn.all_gather(ttnn_tensor, dim=3, num_links=1)

    for device_tensor in ttnn.get_device_tensors(ttnn_tensor):
        device_tensor_torch = ttnn.to_torch(device_tensor)
        assert torch.all(device_tensor_torch == full_tensor)

submesh_devices = mesh_device.get_submeshes((2, 2), ttnn.MeshType.Ring)
for submesh in submesh_devices:
    run_model(submesh)

```

image

### Checklist
- [ ] Post commit CI passes
- [ ] Blackhole Post commit (if applicable)
- [ ] Model regression CI testing passes (if applicable)
- [ ] Device performance regression CI testing passes (if applicable)
- [ ] New/Existing tests provide coverage for changes

### CODE REVIEW

**Review by cglagovichTT — APPROVED** [review]

**Comment by SeanNijjar** [comment]
Hey @cfjchu  I'm mildly concerned with this inversion of responsibility where the ccl must request the topology it "wants" because in general it shouldn't or wouldn't know what it wants.

Instead what I thought the usage model would be is we have a universal entry point for a given ccl op and the ccl op is given APIS to query mesh information so it can set of its a ring or a line. 

Is this still the case except the mesh device infra knows which op variant to dispatch to? Little confused about this interface.

**Comment by cfjchu** [comment]
> Hey @cfjchu I'm mildly concerned with this inversion of responsibility where the ccl must request the topology it "wants" because in general it shouldn't or wouldn't know what it wants.
> 
> Instead what I thought the usage model would be is we have a universal entry point for a given ccl op and the ccl op is given APIS to query mesh information so it can set of its a ring or a line.
> 
> Is this still the case except the mesh device infra knows which op variant to dispatch to? Little confused about this interface.

@SeanNijjar This is a layered problem. So let's talk about what's currently in `main`. Today, our CCL all_gather op, receives a list of devices and then tries to make a decision about whether to use RING / LINE topology based on number of devices (I'm ignoring ttnn.line_all_gather here and just talking about ttnn.all_gather). This "inversion of responsibility" is already happening. 

In `main`, if it infers we should be using ring topology, then we assume that the devices are ordered in a ring order, and today that order is currently tied to the MeshDevice device creation. Part of my changes in this PR is to explicitly decouple the relationship of between how ring-all-gather expects the ordering of devices and how the MeshDevice is instantiated and internally orders the devices. This does not attempt to resolve the existin "inversion of responsibility" that already exists. This PR is needed as a stepping stone so that I can actually attempt to fix the problem. 

My plan is to commonize `ttnn.all_gather(..)` and `ttnn.line_all_gather(..)` into `ttnn.all_gather(..., topology={LINE, RING}` and make the TT-NN user in control of the decision and declare what all-gather topology they want as a function parametrization. This is what we talked about here: https://github.com/tenstorrent/tt-metal/issues/10915. Once we know what topology the user wants, then I think

File: ttnn/cpp/ttnn/operations/ccl/all_gather/device/all_gather_op.cpp [hunk 1]
```diff
@@ -196,17 +197,19 @@ Tensor all_gather(
     if (num_devices == 1){
         topology = all_gather_op::Topology::Linear;
     }
+    auto mesh_device = MeshDevice::fetch_mesh_device(devices);
+    auto ring_devices = mesh_device->get_ring_devices();
```
Comment (SeanNijjar) [inline ttnn/cpp/ttnn/operations/ccl/all_gather/device/all_gather_op.cpp:201]: This line is really my only current concern.

I agree that we should have only one allgather entrypoint which can temporarily be qualified with a topology. Then allgather can request a chip order  accordingly.

However, I wouldn't want the op to have to request a ring or line but instead just be given a sequence of chips and it decides if it can make a ring out of it or not. This way we get portability across galaxy variants ( those that are and aren't torused up) and the op can automatically take advantage of the torus without the model writer having to update all their ccl calls.

This is more inline with allgather forcing a line for a 2 chip invocation.

**Review by SeanNijjar — APPROVED** [review]

File: ttnn/cpp/ttnn/operations/ccl/all_gather/device/all_gather_op.cpp (hunk 1 above)
Comment (cfjchu) [reply ttnn/cpp/ttnn/operations/ccl/all_gather/device/all_gather_op.cpp:201]: > I agree that we should have only one allgather entrypoint which can temporarily be qualified with a topology. Then allgather can request a chip order accordingly. I also agree we need a way to have a consistent way to get an order of chips.

> However, I wouldn't want the op to have to request a ring or line but instead just be given a sequence of chips and it decides if it can make a ring out of it or not.

Seems to conflict. MeshTensor is mapped onto a set of devices, and invocation to `ttnn.all_gather(mesh_tensor)` will land here. What we're doing here is requesting the ring devices from the mesh tensor.

**Review by caixunshiren — APPROVED** [review]

**Review by TT-BrianLiu — APPROVED** [review]
Looks good

**Review by aliuTT — APPROVED** [review]

**Review by ttmchiou — APPROVED** [review]

**Review by kpaigwar — APPROVED** [review]

**Review by johanna-rock-tt — APPROVED** [review]

**Review by tt-rkim — APPROVED** [review]

**Review by dmakoviichuk-tt — APPROVED** [review]
looks good.
But you don't need to add std:: to the size_t

File: ttnn/cpp/ttnn/multi_device.hpp [hunk 2]
```diff
@@ -6,16 +6,25 @@
...
-std::shared_ptr<MeshDevice> open_mesh_device(const MeshShape& mesh_shape, size_t l1_small_size, size_t trace_region_size, size_t num_command_queues, DispatchCoreType dispatch_core_type, const std::pair<size_t, size_t>& offset = {0, 0});
+std::shared_ptr<MeshDevice> open_mesh_device(
+    const MeshShape& mesh_shape,
+    size_t l1_small_size,
+    size_t trace_region_size,
+    size_t num_command_queues,
+    DispatchCoreType dispatch_core_type,
+    MeshType mesh_type = MeshType::RowMajor,
+    const std::pair<size_t, size_t>& offset = std::pair<size_t, size_t>(0, 0),
+    const std::vector<int>& physical_device_ids = std::vector<int>());
```
Comment (ayerofieiev-tt) [inline ttnn/cpp/ttnn/multi_device.hpp:26]: I think you can do this instead
```suggestion
    const std::vector<int>& physical_device_ids = {});
```

File: tt_metal/impl/device/mesh_device.hpp [hunk 3]
```diff
@@ -9,20 +9,39 @@
...
-#include "mesh_device_view.hpp"
 #include "tt_metal/impl/device/device.hpp"
 #include "tt_metal/impl/device/mesh_device_view.hpp"
 
 namespace tt::tt_metal {
 
 using DeviceIds = std::vector<int>;
 using MeshDeviceID = std::size_t;
-using MeshOffset = std::pair<size_t, size_t>;
+using MeshOffset = std::pair<std::size_t, std::size_t>;
```
Comment (ayerofieiev-tt) [inline tt_metal/impl/device/mesh_device.hpp:19]:
```suggestion
using MeshOffset = std::pair<size_t, size_t>;
```

File: tt_metal/impl/device/mesh_device.hpp [hunk 4]
```diff
@@ -72,38 +91,42 @@ class SystemMesh {
...
+    void unmap_mesh_device(const MeshDevice* mesh_device);
+    std::shared_ptr<MeshDevice> get_mesh_device(const std::vector<chip_id_t>& physical_device_ids);
+    Device* get_device(const chip_id_t physical_device_id) const;
 };
 
 class MeshDevice : public std::enable_shared_from_this<MeshDevice> {
+  private:
     MeshDeviceID mesh_id;
     MeshShape mesh_device_shape;
+    MeshType order;
```
Comment (ayerofieiev-tt) [inline tt_metal/impl/device/mesh_device.hpp:115]: so is it Mesh Type or Order?

File: tt_metal/impl/device/mesh_device.hpp [hunk 5]
```diff
@@ -72,38 +91,42 @@ class SystemMesh {
...
     std::shared_ptr<MeshDeviceView> primary_view;
     std::vector<Device *> devices;
-    std::unordered_map<chip_id_t, int> physical_id_to_device_index;
+    std::shared_ptr<MeshDevice> parent_mesh;
+    std::vector<std::shared_ptr<MeshDevice>> submeshes;
```
Comment (ayerofieiev-tt) [inline tt_metal/impl/device/mesh_device.hpp:119]: So who owns who? 
I think one of these links should be weak, maybe parent?

File: tt_metal/impl/device/mesh_device.hpp [hunk 6]
```diff
@@ -72,38 +91,42 @@ class SystemMesh {
...
+        const MeshDeviceConfig &config);
 
     // Unmap MeshDevice, releasing the associated physical devices.
-    void unmap_mesh_device(const std::shared_ptr<MeshDevice> &mesh_device);
+    void unmap_mesh_device(const MeshDevice* mesh_device);
+    std::shared_ptr<MeshDevice> get_mesh_device(const std::vector<chip_id_t>& physical_device_ids);
+    Device* get_device(const chip_id_t physical_device_id) const;
 };
 
 class MeshDevice : public std::enable_shared_from_this<MeshDevice> {
```
Comment (ayerofieiev-tt) [inline tt_metal/impl/device/mesh_device.hpp:111]: for separate, non pr-review discussion:
We are confusing people using same names for things of different meanings.
Device and MeshDevice
MeshDevice does not look like a device (from api standpoint), and can't be used instead of a device.

File: tt_metal/impl/device/mesh_device.hpp (hunk 4 above)
Comment (cfjchu) [reply tt_metal/impl/device/mesh_device.hpp]: MeshType is more suitable. I've fixed this and added some additional comments.

**Comment by cfjchu** [comment]
> looks good. But you don't need to add std:: to the size_t

Thanks for the review. I've stripped the `std::` qualifer

File: ttnn/cpp/ttnn/multi_device.hpp (hunk 2 above)
Comment (cfjchu) [reply ttnn/cpp/ttnn/multi_device.hpp]: done

File: tt_metal/impl/device/mesh_device.hpp (hunk 3 above)
Comment (cfjchu) [reply tt_metal/impl/device/mesh_device.hpp]: done

File: tt_metal/impl/device/mesh_device.hpp (hunk 5 above)
Comment (cfjchu) [reply tt_metal/impl/device/mesh_device.hpp]: parent should be weak. I've changed this to better reflect ownership

File: tt_metal/impl/device/mesh_device.hpp (hunk 6 above)
Comment (cfjchu) [reply tt_metal/impl/device/mesh_device.hpp:111]: We can sync offline if you have further follow-ups. The intent is to have MeshDevice have a Device-like contract/interface as we start integrating this as metal-native
